tests/unit/sharded_copying_manager_tests/common.py

```diff
@@ -0,0 +1,919 @@
+            # we tell it to stop anywhere else.
+            if self._test_required_transition is not None:
+                raise AssertionError(
+                    "Setting new stop state %s with pending required transition %s"
+                    % (stopping_at, self._test_required_transition)
+                )
+            # If we are already in the required_transition_state, consume it.
+            if (
+                self._test_is_stopped
+                and self._test_stop_state == required_transition_state
+            ):
+                self._test_required_transition = None
+            else:
+                self._test_required_transition = required_transition_state
+
+            if self._test_is_stopped and self._test_stop_state == stopping_at:
+                return
+
+            self._test_stop_state = stopping_at
+            self._test_is_stopped = False
+            # This will wake up threads in `_block_if_should_stop_at` which are waiting for a new stopping point.
+            self._test_state_cv.notifyAll()
+
+            start_time = time.time()
+            # Wait until we get to this point.
+            while not self._test_is_stopped:
+                # This will be woken up by the notify in `_block_if_should_stop_at` method.
+                self._test_state_cv_wait_with_timeout(start_time)
+        finally:
+            self._test_state_cv.release()
+
+    def _test_state_cv_wait_with_timeout(self, start_time):
+        """Waits on the `_test_state_cv` condition variable, but will also throw an AssertionError if the wait
+        time exceeded the `start_time` plus `WAIT_TIMEOUT`.
+
+        @param start_time:  The start time when we first began waiting on this condition, in seconds past epoch.
+        @type start_time: Number
+        """
+        deadline = start_time + TestableCopyingManagerThreadedWorker.WAIT_TIMEOUT
+        self._test_state_cv.wait(timeout=(deadline - time.time()) + 0.5)
+        if time.time() > deadline:
+            raise AssertionError(
+                "Deadline exceeded while waiting on condition variable"
+            )
+
+
+class TestableCopyingManagerThreadedWorker(
+    CopyingManagerThreadedWorker, TestableCopyingManagerFlowController
+):
+    """An instrumented version of the CopyingManagerWorker which allows intercepting of requests sent, control when
+    the worker processes new logs, etc.
+
+    This allows for end-to-end testing of the core of the CopyingManagerWorker.
+
+    Doing this right is a bit complicated because the CopyingManagerWorker runs in its own thread.
+
+    Many members of this class are common with the TestableCopyingManager class
+    and mare oved to the base class - TestableCopyingManagerFlowController
+    """
+
+    __test__ = False
+
+    # The different points at which the CopyingManager can be stopped.  See below.
+    SLEEPING = "SLEEPING"
+    SENDING = "SENDING"
+    RESPONDING = "RESPONDING"
```

Comment:
Can't these be inherited from `TestableCopyingManagerFlowController`?  It seems like you have a moderate amount of duplication even though you are inheriting from that class.